include/nbl/video/IGPUCommandPool.h

```diff
@@ -331,8 +334,103 @@ class NBL_API IGPUCommandPool : public core::IReferenceCounted, public IBackendO
     uint32_t m_familyIx;
 
 private:
-    std::atomic_uint32_t m_resetCount = 0;
-    core::CMemoryPool<core::PoolAddressAllocator<uint32_t>, core::default_aligned_allocator, false, uint32_t> m_commandSegmentPool;
+    CCommandSegment* m_segmentLOLHead = nullptr;
+    CCommandSegment* m_segmentLOLTail = nullptr;
+    std::atomic_uint64_t m_resetCount = 0;
+
+    class CCommandSegmentPool
+    {
+    public:
+        CCommandSegmentPool() : m_pool(COMMAND_SEGMENTS_PER_BLOCK*COMMAND_SEGMENT_SIZE, 0u, MAX_COMMAND_SEGMENT_BLOCK_COUNT, MIN_POOL_ALLOC_SIZE) {}
+
+        bool appendToList(CCommandSegment::Iterator& listHeadItr, CCommandSegment*& listTail)
+        {
+            auto segment = m_pool.emplace<CCommandSegment>();
+            if (!segment)
+            {
+                assert(false);
+                return false;
+            }
+
+            if (listTail)
+            {
+                listTail->setNext(segment);
+                listTail = listTail->getNext();
+            }
+            else
+            {
+                assert(!listHeadItr.cmd && !listHeadItr.segment && "List should've been empty.");
+
+                listTail = segment;
+                listHeadItr = segment->begin();
+            }
+
+            return bool(segment);
+        }
+
+        void deleteList(CCommandSegment::Iterator& headItr, CCommandSegment*& tail)
+        {
+            if (tail)
+            {
+                assert(headItr.segment && headItr.cmd);
+
+                auto freeSegment = [this](CCommandSegment* segment)
+                {
+                    if (segment)
+                    {
+                        segment->~CCommandSegment();
+                        m_pool.deallocate(segment, COMMAND_SEGMENT_SIZE);
+                    }
+                };
+
+                CCommandSegment* prevSegment = nullptr;
+                for (auto& itr = headItr; itr != tail->end();)
+                {
+                    prevSegment = itr.segment;
+                    auto prevCmd = *(itr++);
+
+                    prevCmd->~ICommand();
+
+                    // No need to deallocate prevCmd because:
+                    // - it has been allocated from the LinearAddressAllocator where deallocate is a No-OP
+                    // - the memory will get reclaimed in `~LinearAddressAllocator` when whole segment is destroyed
+
+                    if (itr.segment == prevSegment)
+                        continue;
+
+                    // Segment has changed, we need to free the previous one.
+                    freeSegment(prevSegment);
+                }
```

Comment:
`prevSegment` should be assigned to AFTER `freeSegment(prevSegment)` (updated whenever its not equal, essentially) __otherwise the first segment wll NEVER get freed__

sorry that I've made in mistake in my original listing and never re-assigned it https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1040937745

Reply:
nvm I'm dumb, you asssign before you do `++` on the iterator.